// @seamapi/http/lib/seam/connect/routes/access-codes.js
/*
 * Automatically generated by generate-routes.ts.
 * Do not edit this file or add other files to this directory.
 */
import { warnOnInsecureuserIdentifierKey } from '../../../../lib/seam/connect/auth.js';
import { createClient } from '../../../../lib/seam/connect/client.js';
import { isSeamHttpOptionsWithApiKey, isSeamHttpOptionsWithClient, isSeamHttpOptionsWithClientSessionToken, isSeamHttpOptionsWithConsoleSessionToken, isSeamHttpOptionsWithPersonalAccessToken, SeamHttpInvalidOptionsError, } from '../../../../lib/seam/connect/options.js';
import { limitToSeamHttpRequestOptions, parseOptions, } from '../../../../lib/seam/connect/parse-options.js';
import { SeamHttpAccessCodesUnmanaged } from './access-codes-unmanaged.js';
import { SeamHttpClientSessions } from './client-sessions.js';
export class SeamHttpAccessCodes {
    constructor(apiKeyOrOptions = {}) {
        const options = parseOptions(apiKeyOrOptions);
        this.client = 'client' in options ? options.client : createClient(options);
        this.defaults = limitToSeamHttpRequestOptions(options);
    }
    static fromClient(client, options = {}) {
        const constructorOptions = { ...options, client };
        if (!isSeamHttpOptionsWithClient(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing client');
        }
        return new SeamHttpAccessCodes(constructorOptions);
    }
    static fromApiKey(apiKey, options = {}) {
        const constructorOptions = { ...options, apiKey };
        if (!isSeamHttpOptionsWithApiKey(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing apiKey');
        }
        return new SeamHttpAccessCodes(constructorOptions);
    }
    static fromClientSessionToken(clientSessionToken, options = {}) {
        const constructorOptions = { ...options, clientSessionToken };
        if (!isSeamHttpOptionsWithClientSessionToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing clientSessionToken');
        }
        return new SeamHttpAccessCodes(constructorOptions);
    }
    static async fromPublishableKey(publishableKey, userIdentifierKey, options = {}) {
        warnOnInsecureuserIdentifierKey(userIdentifierKey);
        const clientOptions = parseOptions({ ...options, publishableKey });
        if (isSeamHttpOptionsWithClient(clientOptions)) {
            throw new SeamHttpInvalidOptionsError('The client option cannot be used with SeamHttp.fromPublishableKey');
        }
        const client = createClient(clientOptions);
        const clientSessions = SeamHttpClientSessions.fromClient(client);
        const { token } = await clientSessions.getOrCreate({
            user_identifier_key: userIdentifierKey,
        });
        return SeamHttpAccessCodes.fromClientSessionToken(token, options);
    }
    static fromConsoleSessionToken(consoleSessionToken, workspaceId, options = {}) {
        const constructorOptions = { ...options, consoleSessionToken, workspaceId };
        if (!isSeamHttpOptionsWithConsoleSessionToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing consoleSessionToken or workspaceId');
        }
        return new SeamHttpAccessCodes(constructorOptions);
    }
    static fromPersonalAccessToken(personalAccessToken, workspaceId, options = {}) {
        const constructorOptions = { ...options, personalAccessToken, workspaceId };
        if (!isSeamHttpOptionsWithPersonalAccessToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing personalAccessToken or workspaceId');
        }
        return new SeamHttpAccessCodes(constructorOptions);
    }
    get unmanaged() {
        return SeamHttpAccessCodesUnmanaged.fromClient(this.client, this.defaults);
    }
    async create(body) {
        const { data } = await this.client.request({
            url: '/access_codes/create',
            method: 'post',
            data: body,
        });
        return data.access_code;
    }
    async createMultiple(body) {
        const { data } = await this.client.request({
            url: '/access_codes/create_multiple',
            method: 'post',
            data: body,
        });
        return data.access_codes;
    }
    async delete(body) {
        await this.client.request({
            url: '/access_codes/delete',
            method: 'post',
            data: body,
        });
    }
    async generateCode(body) {
        const { data } = await this.client.request({
            url: '/access_codes/generate_code',
            method: 'post',
            data: body,
        });
        return data.generated_code;
    }
    async get(body) {
        const { data } = await this.client.request({
            url: '/access_codes/get',
            method: 'post',
            data: body,
        });
        return data.access_code;
    }
    async list(body) {
        const { data } = await this.client.request({
            url: '/access_codes/list',
            method: 'post',
            data: body,
        });
        return data.access_codes;
    }
    async pullBackupAccessCode(body) {
        const { data } = await this.client.request({
            url: '/access_codes/pull_backup_access_code',
            method: 'post',
            data: body,
        });
        return data.backup_access_code;
    }
    async update(body) {
        await this.client.request({
            url: '/access_codes/update',
            method: 'post',
            data: body,
        });
    }
}
