// @seamapi/http/lib/seam/connect/routes/locks.js
/*
 * Automatically generated by generate-routes.ts.
 * Do not edit this file or add other files to this directory.
 */
import { warnOnInsecureuserIdentifierKey } from '../../../../lib/seam/connect/auth.js';
import { createClient } from '../../../../lib/seam/connect/client.js';
import { isSeamHttpOptionsWithApiKey, isSeamHttpOptionsWithClient, isSeamHttpOptionsWithClientSessionToken, isSeamHttpOptionsWithConsoleSessionToken, isSeamHttpOptionsWithPersonalAccessToken, SeamHttpInvalidOptionsError, } from '../../../../lib/seam/connect/options.js';
import { limitToSeamHttpRequestOptions, parseOptions, } from '../../../../lib/seam/connect/parse-options.js';
import { resolveActionAttempt } from '../../../../lib/seam/connect/resolve-action-attempt.js';
import { SeamHttpActionAttempts } from './action-attempts.js';
import { SeamHttpClientSessions } from './client-sessions.js';
export class SeamHttpLocks {
    constructor(apiKeyOrOptions = {}) {
        const options = parseOptions(apiKeyOrOptions);
        this.client = 'client' in options ? options.client : createClient(options);
        this.defaults = limitToSeamHttpRequestOptions(options);
    }
    static fromClient(client, options = {}) {
        const constructorOptions = { ...options, client };
        if (!isSeamHttpOptionsWithClient(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing client');
        }
        return new SeamHttpLocks(constructorOptions);
    }
    static fromApiKey(apiKey, options = {}) {
        const constructorOptions = { ...options, apiKey };
        if (!isSeamHttpOptionsWithApiKey(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing apiKey');
        }
        return new SeamHttpLocks(constructorOptions);
    }
    static fromClientSessionToken(clientSessionToken, options = {}) {
        const constructorOptions = { ...options, clientSessionToken };
        if (!isSeamHttpOptionsWithClientSessionToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing clientSessionToken');
        }
        return new SeamHttpLocks(constructorOptions);
    }
    static async fromPublishableKey(publishableKey, userIdentifierKey, options = {}) {
        warnOnInsecureuserIdentifierKey(userIdentifierKey);
        const clientOptions = parseOptions({ ...options, publishableKey });
        if (isSeamHttpOptionsWithClient(clientOptions)) {
            throw new SeamHttpInvalidOptionsError('The client option cannot be used with SeamHttp.fromPublishableKey');
        }
        const client = createClient(clientOptions);
        const clientSessions = SeamHttpClientSessions.fromClient(client);
        const { token } = await clientSessions.getOrCreate({
            user_identifier_key: userIdentifierKey,
        });
        return SeamHttpLocks.fromClientSessionToken(token, options);
    }
    static fromConsoleSessionToken(consoleSessionToken, workspaceId, options = {}) {
        const constructorOptions = { ...options, consoleSessionToken, workspaceId };
        if (!isSeamHttpOptionsWithConsoleSessionToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing consoleSessionToken or workspaceId');
        }
        return new SeamHttpLocks(constructorOptions);
    }
    static fromPersonalAccessToken(personalAccessToken, workspaceId, options = {}) {
        const constructorOptions = { ...options, personalAccessToken, workspaceId };
        if (!isSeamHttpOptionsWithPersonalAccessToken(constructorOptions)) {
            throw new SeamHttpInvalidOptionsError('Missing personalAccessToken or workspaceId');
        }
        return new SeamHttpLocks(constructorOptions);
    }
    async get(body) {
        const { data } = await this.client.request({
            url: '/locks/get',
            method: 'post',
            data: body,
        });
        return data.device;
    }
    async list(body) {
        const { data } = await this.client.request({
            url: '/locks/list',
            method: 'post',
            data: body,
        });
        return data.devices;
    }
    async lockDoor(body, options = {}) {
        const { data } = await this.client.request({
            url: '/locks/lock_door',
            method: 'post',
            data: body,
        });
        const waitForActionAttempt = options.waitForActionAttempt ?? this.defaults.waitForActionAttempt;
        if (waitForActionAttempt !== false) {
            return await resolveActionAttempt(data.action_attempt, SeamHttpActionAttempts.fromClient(this.client, {
                ...this.defaults,
                waitForActionAttempt: false,
            }), typeof waitForActionAttempt === 'boolean' ? {} : waitForActionAttempt);
        }
        return data.action_attempt;
    }
    async unlockDoor(body, options = {}) {
        const { data } = await this.client.request({
            url: '/locks/unlock_door',
            method: 'post',
            data: body,
        });
        const waitForActionAttempt = options.waitForActionAttempt ?? this.defaults.waitForActionAttempt;
        if (waitForActionAttempt !== false) {
            return await resolveActionAttempt(data.action_attempt, SeamHttpActionAttempts.fromClient(this.client, {
                ...this.defaults,
                waitForActionAttempt: false,
            }), typeof waitForActionAttempt === 'boolean' ? {} : waitForActionAttempt);
        }
        return data.action_attempt;
    }
}
